### PR TITLE
bpo-44525: Split calls into PRECALL and CALL

### PR DESCRIPTION
**Pull request by markshannon**
Specializing calls is going to need a *lot* of specialized bytecode if we want to specialize all four of `CALL_FUNCTION`, `CALL_FUNCTION_KW`, `CALL_METHOD` and `CALL_METHOD_KW` independently.

Since the only real difference between between the `CALL_FUNCTION` and `CALL_METHOD` forms is shifting the start of the arguments by one or two, it makes sense to move that into a pre-call instruction.
The above four opcodes are replaced by:
* `PRECALL_METHOD`: Pairs with `LOAD_METHOD`. Set the pre-call argument delta, and post-call stack shrink for a method call.
* `CALL_NO_KW`: Combines `CALL_FUNCTION` and `CALL_METHOD`
* `CALL_KW`: Combines `CALL_FUNCTION_KW` and `CALL_METHOD_KW`

This means that we can specialize calls to methods with much extra effort.
Specialization of calls to Python methods happens almost for free (we need an additional check that the number of args is as expected).
Specialization of method descriptors needs a few more instructions, but no extra `ADAPTIVE` form is needed.

[1% speedup](https://gist.github.com/markshannon/8da9c26f06673f12faf56bf3fa1254e3)

I think that the slowdowns are calls to bound-methods. So they are next on the list.

https://bugs.python.org/issue44525

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
This is really exciting work, deltablue, hexiom and mako are pretty CALL_X heavy programs. So the pyperformance results looks promising! I'll try to review within the next week unless someone else beats me to it.

File: Python/ceval.c [hunk 1]
```diff
@@ -4638,6 +4628,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
                         stack_pointer, nargs, kwnames
                     );
                     STACK_SHRINK(postcall_shrink);
+                    postcall_shrink = 1;
```
Comment (Fidget-Spinner) [inline Python/ceval.c]: Note to self: this (and other similar lines below) is done to reset `postcall_shrink` to the default value of 1 for the next eval frame/loop.

File: Python/ceval.c [hunk 2]
```diff
@@ -4678,43 +4670,48 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
 
-        TARGET(CALL_FUNCTION_ADAPTIVE) {
+        TARGET(CALL_NO_KW_ADAPTIVE) {
             SpecializedCacheEntry *cache = GET_CACHE();
-            nargs = cache->adaptive.original_oparg;
+            oparg = cache->adaptive.original_oparg;
             if (cache->adaptive.counter == 0) {
                 next_instr--;
-                if (_Py_Specialize_CallFunction(
+                int nargs = oparg+extra_args;
```
Comment (Fidget-Spinner) [inline Python/ceval.c:4797]: Don't you need to set `extra_args = 0;` after this line too?

Side note: the `extra_args` and `post_args` contracts may be human-error prone for other contributors, since people need to remember to reset them at the end of every call.

File: Python/ceval.c [hunk 3]
```diff
@@ -4741,14 +4739,14 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             goto start_frame;
         }
 
-        TARGET(CALL_FUNCTION_BUILTIN_O) {
+        TARGET(CALL_NO_KW_BUILTIN_O) {
```
Comment (Fidget-Spinner) [inline Python/ceval.c:4859]: Note to self: no need to do the `extra_args` ceremony here because this opcode should never appear with `LOAD_METHOD`.

File: Python/ceval.c [hunk 4]
```diff
@@ -4866,6 +4864,98 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            Py_INCREF(Py_None);
+            STACK_SHRINK(2);
+            SET_TOP(Py_None);
+            Py_DECREF(callable);
+            DISPATCH();
+        }
+
+        TARGET(CALL_NO_KW_METHOD_DESCRIPTOR_O) {
+            assert(_Py_OPCODE(next_instr[-2]) == PRECALL_METHOD);
+            SpecializedCacheEntry *caches = GET_CACHE();
```
Comment (Fidget-Spinner) [inline Python/ceval.c]: GH is complaining this variable is unused (likely in release mode). Maybe merge it with the `assert` below like how you did for `CALL_NO_KW_LIST_APPEND`.

File: Python/ceval.c [hunk 5]
```diff
@@ -4866,6 +4864,98 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
+        TARGET(CALL_NO_KW_LIST_APPEND) {
+            assert(_Py_OPCODE(next_instr[-2]) == PRECALL_METHOD);
+            assert(GET_CACHE()->adaptive.original_oparg == 1);
+            DEOPT_IF(extra_args == 0, CALL_NO_KW);
```
Comment (Fidget-Spinner) [inline Python/ceval.c:4991]: Wow, is it possible to not have a bound method? I'm surprised `!PyList_CheckExact(list)` alone isn't enough.

File: Python/specialize.c [hunk 6]
```diff
@@ -1261,7 +1263,51 @@ specialize_class_call(
     PyObject *callable, _Py_CODEUNIT *instr,
     int nargs, SpecializedCacheEntry *cache)
 {
-    SPECIALIZATION_FAIL(CALL_FUNCTION, SPEC_FAIL_CLASS);
+    SPECIALIZATION_FAIL(CALL_NO_KW, SPEC_FAIL_CLASS);
+    return -1;
+}
+
+static PyMethodDescrObject *_list_append = NULL;
```
Comment (Fidget-Spinner) [inline Python/specialize.c:1270]: Something tells me Victor (and the subinterpreter folks) won't be happy seeing this 😄 . We can probably move it to interpreter state in another PR. WDYT?

File: Python/specialize.c [hunk 7]
```diff
@@ -1261,7 +1263,51 @@ specialize_class_call(
...
+    PyMethodDescrObject *descr, _Py_CODEUNIT *instr,
+    int nargs, SpecializedCacheEntry *cache)
+{
+    int oparg = cache->adaptive.original_oparg;
+    if (nargs - oparg != 1) {
+        SPECIALIZATION_FAIL(CALL_NO_KW, SPEC_FAIL_METHDESCR_NON_METHOD);
+        return -1;
+    }
+    if (_list_append == NULL) {
+        _list_append = (PyMethodDescrObject *)_PyType_LookupId(&PyList_Type, &PyId_append);
```
Comment (Fidget-Spinner) [inline Python/specialize.c:1284]: Side note: If we're going this route to make specialization attempts cheaper, we could consider a unified way to cache these types of lookups. E.g. `len` and `isinstance` specializations need looking into `builtins()` dict every time. Caching the `len` and `isinstance` objects plus the `builtins()` dict tag would make them much cheaper.

**Review by Fidget-Spinner — COMMENTED** [review]
The idea looks good in general. The `dis` docs and their opcodes need eventual updating too. That can probably come in another PR.

File: Python/ceval.c (hunk 2 above)
Comment (markshannon) [reply Python/ceval.c:4797]: We want to leave `extra_args` untouched because we are about to re-execute this instruction (or whatever `_Py_Specialize_CallNoKw` turns it into).

File: Python/ceval.c (hunk 5 above)
Comment (markshannon) [reply Python/ceval.c:4991]: Consider `x.append = list; x.append(range(1))`
We would have a stack (tos to the right): `NULL, list, range(1)` and `extra_args == 0`.

Comment (Fidget-Spinner) [reply Python/ceval.c:4991]: Yikes, thanks for digging that out.

File: Python/specialize.c (hunk 6 above)
Comment (markshannon) [reply Python/specialize.c:1270]: `PyList_Type` is a static global and part of the ABI: https://github.com/python/cpython/blob/main/Misc/stable_abi.txt#L864
So the `list.append` method descriptor might as well be a static global  as well (at least for now).

File: Python/specialize.c (hunk 7 above)
Comment (markshannon) [reply Python/specialize.c:1284]: :+1:

File: Doc/library/dis.rst [hunk 8]
```diff
@@ -1037,13 +1036,11 @@ All of the following opcodes use their arguments.
...
    top.  Below the arguments is a callable object to call.
-   ``CALL_FUNCTION_KW`` pops all arguments and the callable object off the stack,
+   ``CALL_KW`` pops all arguments and the callable object off the stack,
    calls the callable object with those arguments, and pushes the return value
    returned by the callable object.
 
-   .. versionchanged:: 3.6
-      Keyword arguments are packed in a tuple instead of a dictionary,
-      *argc* indicates the total number of arguments.
+   .. versionadded:: 3.1
```
Comment (TeamSpen210) [inline Doc/library/dis.rst]: Typo?

**Review by Fidget-Spinner — APPROVED** [review]
Awesome!

File: Misc/NEWS.d/next/Core and Builtins/2021-12-13-17-12-16.bpo-44525.4-FiSf.rst [hunk 9]
```diff
@@ -0,0 +1,8 @@
+Replace the four call bytecode instructions which one pre-call instruction
+and two call instructions.
+
+Removes ``CALL_FUNCTION``, ``CALL_FUNCTION_KW``, ``CALL_METHOD`` and
+``CALL_METHOD_KW``.
+
+Adds ``CALL_NO_KW`` and ``CALL_KW`` call instructions, and
+``PRECALL_METHOD`` prefix for pairing with ``LOAD_METHOD``.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2021-12-13-17-12-16.bpo-44525.4-FiSf.rst:8]: We've started keeping another log of bytecode changes over in [What's New](https://github.com/python/cpython/blob/main/Doc/whatsnew/3.11.rst#cpython-bytecode-changes). That section should probably be updated with these changes (especially since `CALL_METHOD` and `CALL_METHOD_KW` are mentioned there already).

**Review by brandtbucher — COMMENTED** [review]
Sorry I wasn't able to get to this yesterday (got bogged down in benchmarking issues). Just one fairly minor doc comment that could be followed up on:
